src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs

```diff
@@ -52,65 +63,91 @@ public class SourceRepositoryDependencyProvider : IRemoteDependencyProvider
 
         public bool IsHttp => _sourceRepository.PackageSource.IsHttp;
 
+        public PackageSource Source => _sourceRepository.PackageSource;
+
+        /// <summary>
+        /// Discovers all versions of a package from a source and selects the best match.
+        /// This does not download the package.
+        /// </summary>
         public async Task<LibraryIdentity> FindLibraryAsync(
             LibraryRange libraryRange,
             NuGetFramework targetFramework,
             SourceCacheContext cacheContext,
             ILogger logger,
             CancellationToken cancellationToken)
         {
-            await EnsureResource();
+            AsyncLazy<LibraryIdentity> result = null;
 
-            IEnumerable<NuGetVersion> packageVersions = null;
-            try
-            {
-                if (_throttle != null)
-                {
-                    await _throttle.WaitAsync();
-                }
-                packageVersions = await _findPackagesByIdResource.GetAllVersionsAsync(
-                    libraryRange.Name,
-                    cacheContext,
-                    logger,
-                    cancellationToken);
-            }
-            catch (FatalProtocolException e) when (_ignoreFailedSources)
+            var action = new AsyncLazy<LibraryIdentity>(async () =>
+                await FindLibraryCoreAsync(libraryRange, targetFramework, cacheContext, logger, cancellationToken));
+
+            if (cacheContext.RefreshMemoryCache)
             {
-                if (!_ignoreWarning)
-                {
-                    _logger.LogWarning(e.Message);
-                }
-                return null;
+                result = _libraryMatchCache.AddOrUpdate(libraryRange, action, (k, v) => action);
             }
-            finally
+            else
             {
-                _throttle?.Release();
+                result = _libraryMatchCache.GetOrAdd(libraryRange, action);
             }
 
+            return await result;
```

Comment:
The `result` is always awaited. Therefore it's not necessary to have a complicated cache data structure with `AsyncLazy` as a value.